src/app/games/diam/diam.component.ts

```diff
@@ -0,0 +1,320 @@
+        'm 170.57646,188.51696 v -40.08918 l 75.8509,-52.140047 v 40.086147 z',
+        'm 246.42323,96.285785 v 40.086155 l 101.63847,-0.003 0.001,-40.082256 z',
+        'm 424.66037,148.42345 v 40.08377 l -76.60213,-52.14178 10e-4,-40.082033 z',
+        'm 424.66037,148.42193 v 40.08377 l 7.27884,80.37445 v -40.08377 z',
+        'm 435.26609,466.90582 v -40.08615 l 161.64688,-149.2995 8.9e-4,40.08701 z',
+        'm 166.97969,426.86815 v 40.08638 l 268.28379,-0.0433 v -40.08031 z',
+        'm 2.8324855,277.57297 v 40.08636 L 166.97969,466.94782 v -40.08636 z',
+    ];
+    public selected: Selected | null = null; // TODO: use MGPOptional after strictness merge
+    public lastMoved: LastMoved[] = [];
+
+    public viewInfo: ViewInfo = {
+        boardInfo: [],
+        remainingPieces: [],
+    };
+    constructor(messageDisplayer: MessageDisplayer) {
+        super(messageDisplayer);
+        this.rules = new DiamRules(DiamState);
+        this.availableMinimaxes = [
+            new DiamDummyMinimax(this.rules, 'DiamDummyMinimax'),
+        ];
+        this.encoder = DiamMoveEncoder;
+        this.tutorial = new DiamTutorial().tutorial;
+        this.updateBoard();
+    }
+    public async onCaseClick(x: number): Promise<MGPValidation> {
+        const clickValidity: MGPValidation = this.canUserPlay('#click_' + x);
+        if (clickValidity.isFailure()) {
+            return this.cancelMove(clickValidity.getReason());
+        }
+
+        return this.onCaseClickAfterCheck(x);
+    }
+    private async onCaseClickAfterCheck(x: number): Promise<MGPValidation> {
+        if (this.selected != null) {
+            let move: DiamMove;
+            if (this.selected.type === 'pieceFromReserve') {
+                move = new DiamMoveDrop(x, this.selected.piece);
+            } else {
+                if ((this.selected.position.x + 1) % 8 === x) {
+                    move = new DiamMoveShift(this.selected.position, 'right');
+                } else if ((this.selected.position.x + 7) % 8 === x) {
+                    move = new DiamMoveShift(this.selected.position, 'left');
+                } else {
+                    return this.cancelMove(DiamFailure.MUST_SHIFT_LEFT_OR_RIGHT());
+                }
+            }
+            return this.chooseMove(move, this.rules.node.gameState, null, null);
+        } else {
+            return this.cancelMove(DiamFailure.MUST_SELECT_PIECE_FIRST());
+        }
+    }
+    public async onPieceInGameClick(x: number, y: number): Promise<MGPValidation> {
+        const clickValidity: MGPValidation = this.canUserPlay('#click_' + x + '_' + y);
+        if (clickValidity.isFailure()) {
+            return this.cancelMove(clickValidity.getReason());
+        }
+        if (this.selected != null) {
+            return this.onCaseClickAfterCheck(x);
+        }
+        if (this.rules.node.gameState.getPieceAtXY(x, y).owner === this.rules.node.gameState.getCurrentPlayer()) {
```

Comment:
constantize state ?

Reply:
also, you just created this.getCurrentPlayer :p !

Reply:
ajouté `getState` (et `getPreviousState`) pour éviter de devoir tout le temps dupliquer l'état dans une constante (et risquer de pas le faire au bon moment)